balancer/rls/internal/keys/builder.go

```diff
@@ -0,0 +1,198 @@
+// the data path to build the RLS keys to be used for a given request.
+type BuilderMap map[string]builder
+
+func validateKeyBuilderProto(kbs []*rlspb.GrpcKeyBuilder) error {
+	if len(kbs) == 0 {
+		return errors.New("rls: RouteLookupConfig does not contain any GrpcKeyBuilder")
+	}
+
+	// nameProto is a struct which contains only the fields of interest from
+	// the GrpcKeyBuilder_Name proto. This is to make sure that we don't use
+	// the XXX_* fields (which are part of the struct corresponding to
+	// GrpcKeyBuilder_Name proto in the generated pb.go files) as keys in the
+	// map to determine whether or not we have this Name.
+	type nameProto struct {
+		service string
+		method  string
+	}
+	seenNames := make(map[nameProto]bool)
+	for _, kb := range kbs {
+		names := kb.GetNames()
+		if len(names) == 0 {
+			return fmt.Errorf("rls: GrpcKeyBuilder in RouteLookupConfig does not contain any Name {%+v}", kbs)
+		}
+
+		for _, name := range names {
+			if name.GetService() == "" {
+				return fmt.Errorf("rls: GrpcKeyBuilder in RouteLookupConfig contains a Name field with no Service {%+v}", kbs)
+			}
+			n := nameProto{service: name.GetService(), method: name.GetMethod()}
+			if seenNames[n] {
+				return fmt.Errorf("rls: GrpcKeyBuilder in RouteLookupConfig contains repeated Name field {%+v}", kbs)
+			}
+			seenNames[n] = true
+		}
+
+		seenKeys := make(map[string]bool)
+		for _, matcher := range kb.GetHeaders() {
+			if matcher.GetRequiredMatch() {
+				return fmt.Errorf("rls: GrpcKeyBuilder in RouteLookupConfig has required_match field set {%+v}", kbs)
+			}
+			key := matcher.GetKey()
+			if seenKeys[key] {
+				return fmt.Errorf("rls: GrpcKeyBuilder in RouteLookupConfig contains repeated Key field in headers {%+v}", kbs)
+			}
+			seenKeys[key] = true
+		}
+	}
+	return nil
+}
+
+// MakeBuilderMap parses the provided RouteLookupConfig proto and builds a map
+// of key builders.
+//
+// The following conditions are validated, and an error is returned if any of
+// them is not met:
+// grpc_keybuilders field
+// * must have at least one entry
+// * must not have two entries with the same Name
+// * must not have any entry with a Name with the service field unset or empty
+// * must not have any entries without a Name
+// * must not have a headers entry that has required_match set
+// * must not have two headers entries with the same key within one entry
+func MakeBuilderMap(cfg *rlspb.RouteLookupConfig) (BuilderMap, error) {
+	kbs := cfg.GetGrpcKeybuilders()
+	if err := validateKeyBuilderProto(kbs); err != nil {
+		return nil, err
+	}
+
+	bm := make(map[string]builder)
+	for _, kb := range kbs {
+		var matchers []matcher
+		for _, h := range kb.GetHeaders() {
+			matchers = append(matchers, matcher{key: h.GetKey(), names: h.GetNames()})
+		}
+		b := builder{matchers: matchers}
+		for _, name := range kb.GetNames() {
+			path := "/" + name.GetService() + "/" + name.GetMethod()
```

Comment:
I modelled this after our service config parsing code:
https://github.com/grpc/grpc-go/blob/master/service_config.go#L231

The documentation in rls_config.proto doesn't make it very clear as to whether the service name should include slashes. It does say that it should include the proto package name, which I guess is dot separated.
https://github.com/grpc/grpc-proto/blob/master/grpc/lookup/v1/rls_config.proto#L45

service_config.proto doesn't make it explicit either:
https://github.com/grpc/grpc-proto/blob/master/grpc/service_config/service_config.proto#L38
Although the examples here help.

Do you think it makes more sense to clarify the documentation in rls_config.proto?

Reply:
Re-thinking this, it should be fine for `Service` to contain slashes.  It's just `Method` that must not contain slashes.  It might be worth documenting that in the proto files linked.

Reply:
Thinking about this a little more, it does seem like a challenge to validate this in any form other than saying no-slashes-whatsoever in both Service and Method, or doing absolutely no validation.

Even the gRPC HTTP2 spec is very concise on this:
```
Path → ":path" "/" Service-Name "/" {method name} # But see note below.
Service-Name → {IDL-specific service name}
```

For example, let's say we have entries like this in the provided `grpcKeyBuilder` proto
```
{Service: "foo"},
{Service: "foo/fooo"},
{Service: "foo/fooo", Method: "bar/baar"},
```
and let's say we don't do any validation and add entries to our `keyBuilder` map with the following keys:
```
/foo/
/foo/fooo/
/foo/fooo/bar/baar
``` 

And if `Service: foo` had `Method: fooo`, we would end up with `Path: /foo/fooo` and would match the second entry instead of the first one.

And `Path: /foo/fooo/bar` would correctly match wildcard entry `/foo/fooo/`, but `Path: /foo/fooo/bar/baar2` would not end up matching the same wildcard entry.

But if we are guaranteed that the `Path` in the incoming request would be of the form `/service/method` and would not have any slashes in the service or method parts, could we not get away with no validation on the `grpcKeyBuilder` proto? Basically, this is no different from the MethodConfig in ServiceConfig, right?

Reply:
It shouldn't be unreasonable or problematic for the service name to contain slashes, although it never will in the proto world.  To tear the service name back out of the path we store as the map key, we just remove the leading slash and everything after and including the final slash.  (As a regex: `\/(.*)\/(.*)` - match group 1 is the service and 2 is the method.)

> And if `Service: foo` had `Method: fooo`, we would end up with `Path: /foo/fooo` and would match the second entry instead of the first one.

Why? There is no trailing slash in the request's path (`/foo/fooo`), but there is in the map key (`/foo/fooo/`), so I would expect this to not match.

I agree the same holds true with service configs.  If the method ever contains a slash (anywhere in grpc where we accept a bare method string), things just won't work right.  Whether we choose to validate and give an explicit error or not validate this and have things not work as expected by the user is the question.  I'd rather validate, even though it's unlikely and unusual.

Reply:
Done. Added validation to make sure that method names don't contain slashes. Thanks.